### PR TITLE
[RLlib] DreamerV3: Add Conv2d-transpose support to new model Catalog.

### PR DESCRIPTION
**Pull request by sven1977**
DreamerV3: Add Conv2d-transpose support to new model Catalog.
* primitive conv2d transpose stacks for tf2 and torch
* to match tf2's `Conv2DTranspose` padding="same" setting (which torch does not offer), we had to add a new custom `Stride2D` layer for torch, such that striding can be done before(!) padding.
* Conv2DTransposeHead for tf2 and torch
* test cases comparing the number of params AND 100% matching computations between different configs and frameworks

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/models/configs.py [hunk 1]
```diff
@@ -159,11 +159,11 @@ def build(self, framework: str = "torch") -> Model:
         self._validate(framework=framework)
 
         if framework == "torch":
-            from ray.rllib.core.models.torch.mlp import TorchMLPHead
+            from ray.rllib.core.models.torch.heads import TorchMLPHead
```
Comment (sven1977) [inline rllib/core/models/configs.py:162]: Created a `heads` file (instead of mlp.py). Analogous to the already existing `encoders` file.

File: rllib/core/models/configs.py [hunk 2]
```diff
@@ -246,15 +246,147 @@ def build(self, framework: str = "torch") -> Model:
...
 
             return TorchFreeLogStdMLPHead(self)
         else:
-            from ray.rllib.core.models.tf.mlp import TfFreeLogStdMLPHead
+            from ray.rllib.core.models.tf.heads import TfFreeLogStdMLPHead
 
             return TfFreeLogStdMLPHead(self)
 
 
+@ExperimentalAPI
```
Comment (sven1977) [inline rllib/core/models/configs.py:258]: So now we have two head types (so far):
MLPHead
CNNTransposeHead

File: rllib/core/models/configs.py [hunk 3]
```diff
@@ -246,15 +246,191 @@ def build(self, framework: str = "torch") -> Model:
...
+@dataclass
+class CNNTransposeHeadConfig(ModelConfig):
+    """Configuration for a convolutional transpose head (decoder) network.
+
+    The configured Model transforms 1D-observations into an image space.
+    The stack of layers is composed of an initial Dense layer, followed by a sequence
+    of Conv2DTranspose layers.
+    `input_dims` describes the shape of the (1D) input tensor,
+    `initial_image_dims` describes the input into the first Conv2DTranspose
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/configs.py:267]: If one does not read the full docstring, I think initial_image_dims can be mistaken to be the input dims of the image that is to be reconstructed, so the output dims of the transpose2d.
I think "first_transpose_layer_input_dims" would avoid this.

File: rllib/core/models/configs.py [hunk 4]
```diff
@@ -246,15 +246,191 @@ def build(self, framework: str = "torch") -> Model:
...
+    Example:
+    .. code-block:: python
+        # Configuration:
+        config = CNNTransposeHeadConfig(
+            input_dims=[128],  # 1D input vector (possibly coming from another NN)
+            initial_image_dims=[4, 4, 32],  # first image input to deconv stack
+            cnn_transpose_filter_specifiers=[
+                [16, 4, 2],
+                [3, 4, 2],
+            ],
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/configs.py:379]: Cool, I didn't expect to find this working!

File: rllib/core/models/tests/test_cnn_transpose_heads.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        """Tests building DeConv heads properly and checks for correct architecture."""
+
+        # Loop through different combinations of hyperparameters.
+        inputs_dimss = [[1], [50]]
+        initial_image_dims = [4, 4, 96]
+        cnn_transpose_filter_specifierss = [
+            [[48, 4, 2], [24, 4, 2], [3, 4, 2]],  # double image 3x (32) into an RGB
+            [[48, 4, 2], [24, 4, 2], [1, 4, 2]],  # double image 3x (32) into grayscale
+            [[3, 4, 3]],  # triple image (->12x12) into an RGB
+            [[1, 7, 3]],  # triple image (->12x12) w/ larger filter into grayscale
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/tests/test_cnn_transpose_heads.py:23]: What do "3x (32)"  and "(->12x12)" mean?
Maybe it's me, but I think these comments could be clearer.

File: rllib/core/models/tests/test_cnn_transpose_heads.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            if cnn_transpose_filter_specifiers == [[3, 4, 3]]:
+                expected_output_dims = [12, 12, 3]
+            elif cnn_transpose_filter_specifiers == [[1, 7, 3]]:
+                expected_output_dims = [12, 12, 1]
+            else:
+                expected_output_dims = (
+                    [32, 32, 3]
+                    if cnn_transpose_filter_specifiers[-1][0] == 3
+                    else [32, 32, 1]
+                )
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/tests/test_cnn_transpose_heads.py:54]: Can we include these in the inputs further up and simply pair them with the filters?

File: rllib/core/models/tests/test_cnn_transpose_heads.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+                cnn_transpose_filter_specifiers=cnn_transpose_filter_specifiers,
+                cnn_transpose_activation=cnn_transpose_activation,
+                cnn_transpose_use_layernorm=cnn_transpose_use_layernorm,
+                output_dims=expected_output_dims,
+                use_bias=use_bias,
+            )
+
+            # Use a ModelChecker to compare all added models (different frameworks)
+            # with each other.
+            model_checker = ModelChecker(config)
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/tests/test_cnn_transpose_heads.py:78]: This is just such a great tool! Thanks for coming up with it.

File: rllib/core/models/tests/test_cnn_transpose_heads.py (hunk 5 above)
Comment (sven1977) [reply rllib/core/models/tests/test_cnn_transpose_heads.py:23]: Makes sense, will clarify ..

File: rllib/core/models/tests/test_cnn_transpose_heads.py (hunk 6 above)
Comment (sven1977) [reply rllib/core/models/tests/test_cnn_transpose_heads.py:54]: You are right. This is pretty messy. Will "tuple" them together to make it more structured and generic ...

File: rllib/core/models/tf/heads.py [hunk 8]
```diff
@@ -80,3 +86,54 @@ def _forward(self, inputs: tf.Tensor, **kwargs) -> tf.Tensor:
...
         return logits_out
+
+
+class TfCNNTransposeHead(TfModel):
+    def __init__(self, config: CNNTransposeHeadConfig) -> None:
+        super().__init__(config)
+
+        # Initial, inactivated Dense layer (always w/ bias).
+        # This layer is responsible for getting the incoming tensor into a proper
+        # initial image shape (w x h x filters) for the suceeding Conv2DTranspose stack.
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/tf/heads.py:97]: succeeding

File: rllib/core/models/tf/heads.py [hunk 9]
```diff
@@ -80,3 +86,54 @@ def _forward(self, inputs: tf.Tensor, **kwargs) -> tf.Tensor:
...
+        # Reshape to initial 3D (image-like) format to enter CNN transpose stack.
+        out = tf.reshape(
+            out,
+            shape=(-1,) + tuple(self.config.initial_image_dims),
+        )
+        # Push through CNN transpose stack.
+        out = self.cnn_transpose_net(out)
+        # Add 0.5 to center the (always non-activated, non-normalized) outputs more
+        # around 0.0.
+        return out + 0.5
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/tf/heads.py:139]: Can you expand on this? Maybe also in the comment?

File: rllib/models/torch/misc.py [hunk 10]
```diff
@@ -62,6 +63,96 @@ def same_padding(
...
 
+@DeveloperAPI
+def same_padding_transpose_after_stride(
+    strided_size: Tuple[int, int],
+    kernel: Tuple[int, int],
+    stride: Union[int, Tuple[int, int]],
+) -> (Union[int, Tuple[int, int]], Tuple[int, int]):
+    """Computes padding and output size such that TF Conv2DTranspose `same` is matched.
+
+    Note that when padding="same", tf's Conv2DTranspose makes sure that 0-padding is
```
Comment (ArturNiederfahrenhorst) [inline rllib/models/torch/misc.py:74]: For consistency: _TF's_ ?

File: rllib/models/torch/misc.py [hunk 11]
```diff
@@ -62,6 +63,96 @@ def same_padding(
...
+    pad_top = pad_bottom + (1 if pad_total_h % 2 == 1 else 0)
+
+    # Compute the output size.
+    output_size = (
+        strided_size[0] + pad_total_w - k_w + 1,
+        strided_size[1] + pad_total_h - k_h + 1,
+    )
+
+    # Return padding and output sizes.
+    return (pad_left, pad_right, pad_top, pad_bottom), output_size
```
Comment (ArturNiederfahrenhorst) [inline rllib/models/torch/misc.py:153]: As suggested by your docstrings in this method, `output_size ` is actually output_sizeS or output_shape

**Review by ArturNiederfahrenhorst — APPROVED** [review]
Awesome! Got only a couple of nits.
Thanks for the thorough docstring.

File: rllib/models/torch/misc.py (hunk 11 above)
Comment (sven1977) [reply rllib/models/torch/misc.py:153]: fixed comment

File: rllib/models/torch/misc.py (hunk 10 above)
Comment (sven1977) [reply rllib/models/torch/misc.py:74]: fixed

File: rllib/core/models/tests/test_cnn_transpose_heads.py (hunk 6 above)
Comment (sven1977) [reply rllib/core/models/tests/test_cnn_transpose_heads.py:54]: fixed

File: rllib/core/models/tests/test_cnn_transpose_heads.py (hunk 5 above)
Comment (sven1977) [reply rllib/core/models/tests/test_cnn_transpose_heads.py:23]: fixed
